### PR TITLE
[okina] Some tweaks and fixes in the build system [okina-build-tweaks]

### PR DESCRIPTION
**Pull request by v-dobrev**
TODO:
- [x] Update `INSTALL`
- [x] Consider adding `-L/usr/local/cuda/lib` to `CUDA_LIB`

### CODE REVIEW

File: config/defaults.mk [hunk 1]
```diff
@@ -18,12 +18,16 @@
 # Some choices below are based on the OS type:
 NOTMAC := $(subst Darwin,,$(shell uname -s))
 
-CXX = g++-7
```
Comment (tzanio) [inline config/defaults.mk]: Sorry about that

File: config/defaults.mk [hunk 2]
```diff
@@ -301,45 +300,24 @@ PUMI_OPT = -I$(PUMI_DIR)/include
...
-ifeq ($(MFEM_USE_CUDA),YES)
-   ifndef CUDA_DIR
-      CUDA_DIR := /usr/local/cuda
-   endif
-   CUDA_OPT := -I$(CUDA_DIR)/include
-   CUDA_LIB := -L$(CUDA_DIR)/lib64 -lcuda -lcudart
-endif
-
-# MPI library configuration
-ifeq ($(MFEM_USE_MPI),YES)
```
Comment (tzanio) [inline config/defaults.mk]: This is needed for the parallel build with CUDA, no?

File: config/defaults.mk [hunk 3]
```diff
@@ -301,45 +300,24 @@ PUMI_OPT = -I$(PUMI_DIR)/include
...
-   ifndef MPI_DIR
-      MPI_DIR := $(HOME)/local/openmpi/3.0.0
-   endif
-   MPI_OPT := -I$(MPI_DIR)/include
-   MPI_LIB := -L$(MPI_DIR)/lib -lmpi
-endif
+# CUDA library configuration. Since we compile and link with nvcc (when CUDA is
+# enabled) we only need to explicitly link with the CUDA driver, libcuda.*,
+# which is usually in a system path.
```
Comment (tzanio) [inline config/defaults.mk]: Both `CUDA_DIR` and `MPI_DIR` are currently mentioned in the `INSTALL` file. If we don't need them (which will be great) then this PR should also update `INSTALL`

File: config/defaults.mk (hunk 2 above)
Comment (v-dobrev) [reply config/defaults.mk]: It turns out, you can just tell `nvcc` to use the mpi wrapper and that seems to work. Still need to test that on other systems though.

File: config/defaults.mk (hunk 3 above)
Comment (v-dobrev) [reply config/defaults.mk]: Good point.

File: config/defaults.mk (hunk 2 above)
Comment (tzanio) [reply config/defaults.mk]: > you can just tell nvcc to use the mpi wrapper and that seems to work

That will be ideal, if it works on LC, Linux and Mac? Should we document it in `INSTALL`?

**Comment by camierjs** [comment]
Thank you for this PR @v-dobrev 
We should add the `XCOMPILER` before the `-fopenmp` for `m cuda MFEM_USE_OPENMP=YES`.
Do you have a `-lcuda` on the mac? I installed the driver, but building the examples still fails.
Is it because it's `libcuda.dylib`?

File: config/defaults.mk [hunk 4]
```diff
@@ -301,45 +300,24 @@ PUMI_OPT = -I$(PUMI_DIR)/include
...
+CUDA_OPT =
+CUDA_LIB = -lcuda
```
Comment (tzanio) [inline config/defaults.mk]: I need `-L /usr/local/cuda/lib` on my Mac

**Comment by tzanio** [comment]
Approved modulus TODO items

File: config/defaults.mk [hunk 5]
```diff
@@ -35,45 +39,40 @@ STATIC = YES
...
-ifeq ($(MFEM_USE_CUDA),YES)
-   # Pass the following arguments to the host compiler
-   MFEM_XARCHIVE  = -Xarchive
-   MFEM_XCOMPILER = -Xcompiler
-else
-   MFEM_XARCHIVE  =
-   MFEM_XCOMPILER =
-endif
+CUDA_CXX = nvcc
+CUDA_ARCH = sm_60
```
Comment (tzanio) [inline config/defaults.mk]: This should also be documented in the `INSTALL` file.

**Comment by v-dobrev** [comment]
@tzanio, @camierjs, does either of you have nvidia gpu on your macs? On such a system, the cuda driver will probably be installed in a system folder along with the graphics drivers. We should figure this out before settling on a solution. Another point is that `nvcc` actually adds the `lib/lib64` directory from its installation with an `-L` flag (you can see that if you run `nvcc` with the `-v` option). The failure you see probably means that `-lcuda` appears before that `-L` flag.

File: general/device.cpp [hunk 6]
```diff
@@ -25,12 +25,12 @@ OccaDevice occaDevice;
 static void DeviceSetup(const int dev, int &ngpu)
 {
    cudaGetDeviceCount(&ngpu);
-   MFEM_ASSERT(ngpu>0, "No CUDA device found!");
+   if (!(ngpu>0)) { mfem_error("No CUDA device found!"); }
```
Comment (v-dobrev) [inline general/device.cpp]: `MFEM_VERIFY` is a more natural choice probably.

**Comment by camierjs** [comment]
We dont have NVIDIA GPUs, yes.
The toolkit installed thigs in `/Developer/NVIDIA/CUDA-10.1/lib` and also `/usr/local/cuda`:
```lrwxr-xr-x@  1 root  wheel    58 Feb  9 09:23 CUDA_Toolkit_Release_Notes.txt -> /Developer/NVIDIA/CUDA-10.1/CUDA_Toolkit_Release_Notes.txt
lrwxr-xr-x@  1 root  wheel    36 Feb  9 09:23 EULA.txt -> /Developer/NVIDIA/CUDA-10.1/EULA.txt
lrwxr-xr-x@  1 root  wheel    48 Feb  9 09:23 NsightCompute-2019.1 -> /Developer/NVIDIA/CUDA-10.1/NsightCompute-2019.1
lrwxr-xr-x@  1 root  wheel    48 Feb  9 09:23 NsightSystems-2018.3 -> /Developer/NVIDIA/CUDA-10.1/NsightSystems-2018.3
lrwxr-xr-x@  1 root  wheel    31 Feb  9 09:23 bin -> /Developer/NVIDIA/CUDA-10.1/bin
lrwxr-xr-x@  1 root  wheel    31 Feb  9 09:23 doc -> /Developer/NVIDIA/CUDA-10.1/doc
lrwxr-xr-x@  1 root  wheel    34 Feb  9 09:23 extras -> /Developer/NVIDIA/CUDA-10.1/extras
lrwxr-xr-x@  1 root  wheel    35 Feb  9 09:23 include -> /Developer/NVIDIA/CUDA-10.1/include
lrwxr-xr-x@  1 root  wheel    31 Feb  9 09:23 jre -> /Developer/NVIDIA/CUDA-10.1/jre
drwxr-xr-x  82 root  wheel  2624 Apr  6 16:28 lib
lrwxr-xr-x@  1 root  wheel    37 Feb  9 09:23 libnsight -> /Developer/NVIDIA/CUDA-10.1/libnsight
lrwxr-xr-x@  1 root  wheel    35 Feb  9 09:23 libnvvp -> /Developer/NVIDIA/CUDA-10.1/libnvvp
lrwxr-xr-x@  1 root  wheel    44 Feb  9 09:23 nsightee_plugins -> /Developer/NVIDIA/CUDA-10.1/nsightee_plugins
lrwxr-xr-x@  1 root  wheel    32 Feb  9 09:23 nvvm -> /Developer/NVIDIA/CUDA-10.1/nvvm
lrwxr-xr-x@  1 root  wheel    35 Feb  9 09:23 samples -> /Developer/NVIDIA/CUDA-10.1/samples
lrwxr-xr-x@  1 root  wheel    31 Feb  9 09:23 src -> /Developer/NVIDIA/CUDA-10.1/src
lrwxr-xr-x@  1 root  wheel    33 Feb  9 09:23 tools -> /Developer/NVIDIA/CUDA-10.1/tools
lrwxr-xr-x@  1 root  wheel    39 Feb  9 09:23 version.txt -> /Developer/NVIDIA/CUDA-10.1/version.txt```

Only `libcuda.dylib` is not a link inside the `/usr/local/cuda/lib`

**Comment by camierjs** [comment]
`-L/usr/local/cuda/lib` works fine for me now, thank you @tzanio & @v-dobrev

**Pull request merged**